pkg/fatfs/fatfs_system/ffsystem.c

```diff
@@ -0,0 +1,97 @@
+void *ff_memalloc(  /* Returns pointer to the allocated memory block (null if not enough core) */
+    UINT msize      /* Number of bytes to allocate */
+    )
+{
+    return malloc(msize);   /* Allocate a new memory block with POSIX API */
+}
+
+/*------------------------------------------------------------------------*/
+/* Free a memory block                                                    */
+/*------------------------------------------------------------------------*/
+void ff_memfree(
+    void *mblock    /* Pointer to the memory block to free (nothing to do if null) */
+    )
+{
+    free(mblock);   /* Free the memory block with POSIX API */
+}
+#endif
+
+#if FF_FS_REENTRANT /* Mutal exclusion */
+static mutex_t mutex[FF_VOLUMES];
+
+/*------------------------------------------------------------------------*/
+/* Create a Synchronization Object                                        */
+/*------------------------------------------------------------------------*/
+/* This function is called in f_mount() function to create a new
+ * synchronization object for the volume, such as semaphore and mutex.
+ * When a 0 is returned, the f_mount() function fails with FR_INT_ERR.
+ */
+int ff_cre_syncobj(     /* 1:Function succeeded, 0:Could not create the sync object */
+    BYTE vol,           /* Corresponding volume (logical drive number) */
+    FF_SYNC_t *sobj     /* Pointer to return the created sync object */
+    )
+{
+    *sobj = &mutex[vol];
+    mutex_init(*sobj);
+
+    return true;
+}
+
+/*------------------------------------------------------------------------*/
+/* Delete a Synchronization Object                                        */
+/*------------------------------------------------------------------------*/
+/* This function is called in f_mount() function to delete a synchronization
+ * object that created with ff_cre_syncobj() function. When a 0 is returned,
+ * the f_mount() function fails with FR_INT_ERR.
+ */
+int ff_del_syncobj(     /* 1:Function succeeded, 0:Could not delete due to an error */
+    FF_SYNC_t sobj      /* Sync object tied to the logical drive to be deleted */
+    )
+{
+    (void)sobj;
+    return true;
+}
+
+/*------------------------------------------------------------------------*/
+/* Request Grant to Access the Volume                                     */
+/*------------------------------------------------------------------------*/
+/* This function is called on entering file functions to lock the volume.
+ * When a 0 is returned, the file function fails with FR_TIMEOUT.
+ */
+int ff_req_grant(   /* 1:Got a grant to access the volume, 0:Could not get a grant */
+    FF_SYNC_t sobj  /* Sync object to wait */
+    )
+{
+    if (IS_USED(MODULE_ZTIMER_MSEC)) {
+        return ztimer_mutex_lock_timeout(ZTIMER_MSEC, sobj, FF_FS_TIMEOUT) == 0;
+    }
+    else {
+        return ztimer_mutex_lock_timeout(ZTIMER_USEC, sobj, US_PER_MS * FF_FS_TIMEOUT) == 0;
+    }
```

Comment:
I'd rather pull-in `ztimer_msec` as dependency.

I think the assumption `ztimer_usec` is around otherwise is valid for most cases. But I guess you could create one case where it's not available.

Reply:
Hm my idea was that I would not have to add an explicit ztimer dependency to fatfs and it would just use whatever is there already.
Then I could use it in riotboot and set `FATFS_FFCONF_OPT_FS_REENTRANT` to 0 without pulling in a timer.

Reply:
Maybe model `FATFS_FFCONF_OPT_FS_REENTRANT` as pseudo-module? This pseudo-module requires `ztimer`. RIOTBOOT is single-threaded and shouldn't pull-in the reentrant part of FatFS.

Reply:
Ok, I just wanted to be lazy :wink:

Reply:
TBH this looks much cleaner and is less surprising for first-time users :-)

Reply:
The only drawback is that it doesn't work :(

Reply:
Works when I add `fatfs_threadsafe` to `USEMODULE` instead of `DEFAULT_MODULE`.